DistroLauncher/OOBE.cpp

```diff
@@ -33,81 +35,43 @@ namespace DistributionInfo {
 		return ((SUCCEEDED(hr)) && (exitCode == 0));
 	}
 
-	ULONG DistributionInfo::OOBE()
+	HRESULT DistributionInfo::OOBESetup()
 	{
-		ULONG uid = UID_INVALID;
-
-		// calling the oobe experience
-		DWORD exitCode;
 		// Prepare prefill information to send to the OOBE.
 		std::wstring prefillCLIPostFix = DistributionInfo::PreparePrefillInfo();
 		std::wstring commandLine = DistributionInfo::OOBE_NAME + prefillCLIPostFix;
+		// calling the OOBE.
+		DWORD exitCode=-1;
 		HRESULT hr = g_wslApi.WslLaunchInteractive(commandLine.c_str(), true, &exitCode);
 		if ((FAILED(hr)) || (exitCode != 0)) {
-			return uid;
+			return hr;
 		}
 
 		hr = g_wslApi.WslLaunchInteractive(L"clear", true, &exitCode);
 		if (FAILED(hr)) {
-			return uid;
+			return hr;
 		}
 
-		// getting username from ouput
-		// Create a pipe to read the output of the launched process.
-		HANDLE readPipe;
-		HANDLE writePipe;
-		SECURITY_ATTRIBUTES sa{ sizeof(sa), nullptr, true };
-		if (CreatePipe(&readPipe, &writePipe, &sa, 0)) {
-			// Query the UID of the supplied username.
-			std::wstring command = L"cat /var/lib/ubuntu-wsl/assigned_account";
-			int returnValue = 0;
-			HANDLE child;
-			HRESULT hr = g_wslApi.WslLaunch(command.c_str(), true, GetStdHandle(STD_INPUT_HANDLE), writePipe, GetStdHandle(STD_ERROR_HANDLE), &child);
-			if (SUCCEEDED(hr)) {
-				// Wait for the child to exit and ensure process exited successfully.
-				WaitForSingleObject(child, INFINITE);
-				DWORD exitCode;
-				if ((GetExitCodeProcess(child, &exitCode) == false) || (exitCode != 0)) {
-					hr = E_INVALIDARG;
-				}
-
-				CloseHandle(child);
-				if (SUCCEEDED(hr)) {
-					char buffer[64];
-					DWORD bytesRead;
-
-					// Read the output of the command from the pipe and query UID
-					if (ReadFile(readPipe, buffer, (sizeof(buffer) - 1), &bytesRead, nullptr)) {
-						buffer[bytesRead] = ANSI_NULL;
-						try {
-							const size_t bfrSize = strlen(buffer) + 1;
-							wchar_t* wbuffer = new wchar_t[bfrSize];
-							size_t outSize;
-							mbstowcs_s(&outSize, wbuffer, bfrSize, buffer, bfrSize - 1);
-							std::wstring_view uname_bfr{ wbuffer, bfrSize };
-							uid = QueryUid(uname_bfr);
-						}
-						catch (...) {}
-					}
-				}
-			}
-
-			CloseHandle(readPipe);
-			CloseHandle(writePipe);
+		// read the OOBE exit status file.
+		// Even without interop activated Windows can still access Linux files under WSL.
+		const TCHAR* launcherStatusPath = L"/run/subiquity/launcher-status";
```

Comment:
I would move this file to something else, as we may reuse it for other things to tell the launcher to reboot/shutdown.

So maybe `/run/launcher-command` directly or any better name?

Reply:
That's fine. We just need to update Subiquity accordingly. If we go with my other suggestion, on the default UID, I'd like to have them close, in the same directory, but separate to make things simple. Of course we could read the /etc/wsl.conf generated by the OOBE, find the username and call this function (or do what it does: launch `id -u`, create a pipe to read the stdout and register the uid as the default): 

https://github.com/ubuntu/WSL/blob/dd78a1acf50a66e4131614aa116c11f73c72b5c8/DistroLauncher/DistroLauncher.cpp#L61

Two files of one word each seems simpler and straightforward to me. If you believe that could be somewhat error prone due lack of validation, we can move to some standard syntax, like yaml or ini and put both pieces of information on the same file.

What do you think?

Reply:
> Of course we could read the /etc/wsl.conf generated by the OOBE,
I think if this is not useful (which is not as we will use the API call), let’s not add it to `wsl.conf` anymore.

So yeah, let’s do one file maybe? If you prefer yaml or json, or just a keyfile, this is fine. Maybe one with `action:` and one for extra info like `user`. That way, that file becomes the communication handler between the rootfs (distro) and the laucher (windows)?